src/Execution/BatchLoader/RelationBatchLoader.php

```diff
@@ -53,6 +53,7 @@ public function load(Model $model): Deferred
     {
         $modelKey = ModelKey::build($model);
         $this->parents[$modelKey] = $model;
+        $this->hasResolved = false;
```

Comment:
I am confused why this is necessary, can you explain?

Reply:
The first request without retrieving anything from the cache execution is like this:

1. posts
2. creating posts BatchLoader
3. resolving posts BatchLoader
4. comments 
5. creating comments BatchLoader 
6. resolving comments BatchLoader

The second request with retrieving from the cache:

1. posts
2. creating comments BatchLoader
3. creating posts BatchLoader
4. resolving comments BatchLoader
5. resolving posts BatchLoader
6. comments 
7. adding new keys to the current resolved comments BatchLoader

I know that this is not the RelationBatchLoader problem, but by adding this line I force the batch loader to be run again if there are new keys inside the batch loader.

Reply:
I don't have any idea why a comment batch loader must be created when we are in the step to resolve the posts.

Reply:
I think this fix leads to suboptimal duplicate queries - here is the output of the debugging statements I added:

```
"First request" // /workdir/tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:630
"select * from `posts` where `posts`.`deleted_at` is null" // /workdir/tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:632
"select * from `comments` where `comments`.`post_id` in (1, 2, 3, 4, 5, 6, 7, 8)" // /workdir/tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:632
"select * from `users` where `users`.`id` in (1, 9)" // /workdir/tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:632
"Second request" // /workdir/tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:651
"select * from `users` where `users`.`id` in (1)" // /workdir/tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:632
"select * from `comments` where `comments`.`post_id` in (5)" // /workdir/tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:632
"select * from `users` where `users`.`id` in (1, 9)" // /workdir/tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:632
```

Reply:
Do you have any idea why when cache directive is used, the order of field resolving is changed?

Reply:
I figured out it was due to the relation batch loader returning a `Deferred` result, whereas retrieving a value from cache returns a result immediately. To smooth things over, I tried returning a `Deferred` from `CacheDirective` and found that it fixes the duplicate request for `users` and makes this fix unnecessary.

```
"First request" // /workdir/tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:630
"select * from `posts` where `posts`.`deleted_at` is null" // /workdir/tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:632
"select * from `comments` where `comments`.`post_id` in (1, 2, 3, 4, 5, 6, 7, 8)" // /workdir/tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:632
"select * from `users` where `users`.`id` in (1, 9)" // /workdir/tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:632
"Second request" // /workdir/tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:651
"select * from `comments` where `comments`.`post_id` in (5)" // /workdir/tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:632
"select * from `users` where `users`.`id` in (1, 9)" // /workdir/tests/Integration/Execution/DataLoader/RelationBatchLoaderTest.php:632
```